src/LandIce/evaluators/LandIce_HydrostaticPressure.hpp

```diff
@@ -59,7 +59,17 @@ class HydrostaticPressure : public PHX::EvaluatorWithBaseImpl<Traits>,
   double rho_i;  // [kg m^{-3}]
   double g;      // [m s^{-2}]
 
+  const double pow3 = 1.0e3;
+
   PHAL::MDFieldMemoizer<Traits> memoizer;
+
+  typedef Kokkos::View<int***, PHX::Device>::execution_space ExecutionSpace;
+
+  typedef Kokkos::MDRangePolicy< ExecutionSpace, Kokkos::Rank<2> > Hydrostatic_Pressure_Policy;
```

Comment:
So for this kernel the latter you mentioned is exactly what we want to do. I was hoping the MDRange policy provides this without having to worry about retrieving indices (i.e. adding code bloat). My understanding was that the MDRange policy was supposed to only be used for tightly nested parallelism like this one. We've been running some tests and it doesn't look like it's working as I hoped so we may have to just use a regular RangePolicy.

We'll have to edit the view layout for other, more complex kernels to improve data access patterns. The layout should be fine for this one.

Reply:
I did a timing comparison for MDRange and Range and got an interesting result. The reason I have MDRange policies in place with the node/qp index first is that for tetrahedral elements, there is a pretty significant speedup for some evaluators. 

Here is one example from the comparison on tets:

MDRange (cells, nodes/qp) Phalanx: Evaluator 62: [\<Residual\>] Phi: 0.13328 - 6.92016% [101]
MDRange (nodes/qp, cells) Phalanx: Evaluator 62: [\<Residual\>] Phi: 0.00465843 - 0.246764% [101]
Range   (cells)                      Phalanx: Evaluator 62: [\<Residual\>] Phi: 0.0891155 - 4.89937% [101]

However, after running the comparison on wedge elements, as expected, the simple range policy over cells is the clear better performer.

Reply:
What about Range(cells*nodes) or Range(cells*qp)?

Also, the way you order in MDRange clearly impacts the performance, since it affects how reads will be (coalesced or not) on GPU. I _thought_ Phalanx uses layout left on Cuda, but your timings make me wonder whether it was always Right instead. I have to check.

Also, if you try Range(cells*nodes), you should try to retrieve icell and inode both ways:

```
icell = iter / numNodes;
inode = iter % numNodes;
```
and
```
inode = iter / numCells;
icell = iter % numCells;
```
One of these will be significantly faster than the other. What I would like to know, however, is how these compare to the corresponding MDRange. Since our tiles are 1x1, if I recall correctly a conversation with a kokkos developer, then MDRange might not be beneficial (and even hurt a bit). But I might be wrong. Worth checking, if you have time.

Reply:
Max, are you saying Range(cell) is performing better than MDRange(cell,node) and MDRange(node, cell) on wedge elements?

The fact that the times for the MDRange(node, cell) for both residual and jacobian on tets are the same seems extremely suspicious to me. Just want to confirm, you have `export CUDA_LAUNCH_BLOCKING=1` in your script, correct?

Luca, FYI, these are timings for a different evaluator, `LandIce_LiquidWaterFraction_Def.hpp`. There's some possible thread divergence there so memory accesses may not be coalesced. Phalanx does use left layout (as far as I know unless things have changed recently). These MDRange policies are giving us different results depending on the element type (for wedges, it seems to be working correctly on a simple kernel (i.e. (cell, node) is faster) but not for tets). What you suggest is worth trying to see how it compares with the MDRange policy but maybe on the simple kernel (i.e. HydrostaticPressure). If `icell` as the faster index doesn't perform better in that experiment, I would not be happy. If we go that route though, we'll have to create a layer of abstraction to deal with different data layouts which I was hoping to avoid by using MDRange policies.

If what you're saying is correct about MDRange policy, I'm not sure when they would ever be useful. My understanding was that it was for tightly nested loops and team policies were for more complex nested loops. I could be wrong too though and these things have been changing so it's probably better to ask a kokkos dev and test it ourselves.

My suggestion would be to just go with the Range(cell) policy for now and we can come back to this later. Max, I have some code/scripts we could use to test all these cases out in an isolated environment, that might be more efficient.

Reply:
Wait, why would you prefer Range(numCells) over Range(numCells*numNodes) or Range(numCells*numQPs)? Assuming fully independent nodes/QPs (which is usually the case, if not always), you would be leaving, for nodes, a factor 4 (or more for hexa/wedges) on the table (qps it depends, for stokesFo nothing, since there's 1 qp only).

Obv, one needs to carefully do the math with indices, since you want the fast striding thread idx to match the fast striding idx in the view layout. But if you do that, then you gain quite a bit.

As for MDRanges, we should indeed talk to kokkos devs. My impression is based on a memory of a conversation I had in late 2016, so I may have completely made up stuff. One thing that could explain the MDRange behavior is if internally kokkos maps, say, 1st index to blocks (or warp) and 2nd index to warp (or warp lane). Different elements yield different blocking structures (or, morally, different internal "team" policies), which may utilize the gpu in very different ways. So going from 4 nodes/elem to 6 nodes/elem might screw up gpu occupancy, leaving a lot of stuff on the table. Beware, this is a very wild guess.

Reply:
I prefer Range(numCells) **for now** because it's often enough for full occupancy and these kernels are a very small portion of the overall runtime so it won't make a different in the short term. Range(numCells*numNodes) requires an abstraction for different layouts which can be done but it would take more time and would have to be maintained. I'd like to avoid that with the MDRange policies. If that turns out to be less performant than Range(numCells*numNodes) than we can create the abstraction but I really think that should be something kokkos should be able to handle.

You only gain quite a bit when there's not enough cells to fill the gpu (i.e. strong scaling). In some cases, you actually lose performance, for example, when threading over nodes can affect coalesced memory access.

There's a kokkos training next week, maybe we can get some of our questions answered. Even so, I'd still like to see some isolated evidence of a Range(numCells*numNodes) policy performing better than an MDRange policy across different architectures to send over to the kokkos team. I think I did this before with Aeras but that was 3 or 4 years ago. We need some updated data because I think there's been some changes to the MDRange policy.

Reply:
Thanks, now I see your point. Then it's ok, we can keep Range(numCells) if it works better than the MDRange.

Reply:
We'll keep you posted. Hopefully we can transition to something that works better all around soon.